packages/ListBox/src/components/Options/helpers/options.js

```diff
@@ -0,0 +1,234 @@
+}
+
+export function isOptionVisible(state, key) {
+  if (!Object.keys(state.options).length) return;
+
+  const keyInt = Number.parseInt(key, 10);
+
+  if (typeof state.options[keyInt] === "undefined") {
+    return true;
+  }
+
+  if (state.noResultsFound) {
+    return false;
+  }
+
+  if (state.options[keyInt].isHidden) {
+    return false;
+  }
+
+  return !state.filteredOptions.length || state.filteredOptions.includes(keyInt);
+}
+
+function isDisabled(state, key) {
+  return state.options[key].isDisabled;
+}
+
+export function getNextOptionActiveIndex(state, isAscending = true) {
+  const { activeOption, filteredOptions, options } = state;
+  const optionsKeys = Object.keys(state.options);
+
+  if (state.noResultsFound) return null;
+
+  if (state.activeOption === null) return 0;
+
+  if (optionsKeys.length === 1) {
+    return 0;
+  }
+
+  if (optionsKeys.length - 1 < activeOption) {
+    return 0;
+  }
+
+  let key = state.options[activeOption].index;
+
+  let index = null;
+  let keepIterating = true;
+
+  while (keepIterating) {
+    if (isAscending) {
+      if (filteredOptions.length === 1) {
+        keepIterating = false;
+        return filteredOptions[0];
+      }
+
+      if ((filteredOptions.length && key + 1 > options.length) || key + 1 > optionsKeys.length - 1) {
+        keepIterating = false;
+        return null;
+      }
+
+      key++;
+    } else {
+      if ((filteredOptions.length && key - 1 < 0) || key - 1 < 0) {
+        keepIterating = false;
+        return null;
+      }
+
+      key--;
+    }
+
+    if (isOptionVisible(state, key) && !isDisabled(state, key)) {
+      index = key;
+      keepIterating = false;
+    }
+  }
+
+  return index || 0;
+}
+
+export function getNextOptionActiveIndexLooping(state) {
+  let activeIndex = getNextOptionActiveIndex(state);
+  if (!activeIndex) {
+    activeIndex = getNextOptionActiveIndex(state, false);
+  }
+
+  return activeIndex;
+}
+
+export function handleArrowKeys({ event, state, dispatch, isArrowDown = null }) {
+  if (!state.isOpen) {
+    dispatch({ type: useListBox.types.openPopover });
+    return;
+  }
+
+  event.preventDefault();
+  const next = isArrowDown ? getNextOptionActiveIndex(state) : getNextOptionActiveIndex(state, false);
+  if (next !== null) {
+    if (state.isMulti) {
+      dispatch({
+        type: useListBox.types.setActiveOption,
+        payload: { activeOptionIndex: next, isOpen: true },
+      });
+    } else {
+      selectSingleOption({ activeOptionIndex: next, isOpen: true, state, dispatch });
+    }
+  }
+}
+
+export const handleClickOption = ({ props, state, dispatch }) => event => {
+  const { index } = props;
+  const { options, hasFilter, isMulti, refFilterInput } = state;
+  const hasPreventDefaultOnSelect = options[index].preventDefaultOnSelect;
+  if (state.isDisabled || props.isDisabled) {
+    return;
+  }
+
+  if (state.refListBox.current.contains(event.target) && document.activeElement === document.body && !hasFilter) {
```

Comment:
Wha? Maybe...

```js
const isTheImportantThingTrue = state.refListBox.current.contains(event.target) && document.activeElement === document.body && !hasFilter;
if (isTheImportantThingTrue) state.refListBoxContainer.current.focus();
```

...would give you the opportunity to self-document this condition.